services/brig/test/integration/API/OAuth.hs

```diff
@@ -0,0 +1,865 @@
+
+testRefreshAccessTokenAccessDeniedWhenDisabled :: Opt.Opts -> Brig -> Http ()
+testRefreshAccessTokenAccessDeniedWhenDisabled opts brig = do
+  uid <- randomId
+  let redirectUrl = mkUrl "https://example.com"
+  let scopes = OAuthScopes $ Set.fromList [ReadSelf]
+  (cid, secret, code) <- generateOAuthClientAndAuthorizationCode brig uid scopes redirectUrl
+  let accessTokenRequest = OAuthAccessTokenRequest OAuthGrantTypeAuthorizationCode cid secret code redirectUrl
+  accessToken <- createOAuthAccessToken brig accessTokenRequest
+  withSettingsOverrides (opts & Opt.optionSettings . Opt.oauthEnabledInternal ?~ False) $ do
+    let refreshAccessTokenRequest = OAuthRefreshAccessTokenRequest OAuthGrantTypeRefreshToken cid secret (oatRefreshToken accessToken)
+    refreshOAuthAccessToken' brig refreshAccessTokenRequest !!! assertAccessDenied
+
+testRegisterOAuthClientAccessDeniedWhenDisabled :: Opt.Opts -> Brig -> Http ()
+testRegisterOAuthClientAccessDeniedWhenDisabled opts brig =
+  withSettingsOverrides (opts & Opt.optionSettings . Opt.oauthEnabledInternal ?~ False) $ do
+    let newOAuthClient = newOAuthClientRequestBody "E Corp" "https://example.com"
+    registerNewOAuthClient' brig newOAuthClient !!! assertAccessDenied
+
+assertAccessDenied :: Assertions ()
+assertAccessDenied = do
+  const 403 === statusCode
+  const (Just "forbidden") === fmap Error.label . responseJsonMaybe
+
+testAccessResourceSuccessNginz :: Brig -> Nginz -> Http ()
+testAccessResourceSuccessNginz brig nginz = do
+  -- with ZAuth header
+  user <- createUser "alice" brig
+  let email = fromMaybe (error "no email") $ userEmail user
+  zauthToken <- decodeToken <$> (login nginz (defEmailLogin email) PersistentCookie <!! const 200 === statusCode)
+  get (nginz . path "/self" . header "Authorization" ("Bearer " <> toByteString' zauthToken)) !!! const 200 === statusCode
+
+  -- with Authorization header containing an OAuth bearer token
+  let redirectUrl = mkUrl "https://example.com"
+  let scopes = OAuthScopes $ Set.fromList [ReadSelf]
+  (cid, secret, code) <- generateOAuthClientAndAuthorizationCode brig (userId user) scopes redirectUrl
+  let accessTokenRequest = OAuthAccessTokenRequest OAuthGrantTypeAuthorizationCode cid secret code redirectUrl
+  oauthToken <- oatAccessToken <$> createOAuthAccessToken brig accessTokenRequest
+  get (nginz . paths ["self"] . authHeader oauthToken) !!! const 200 === statusCode
+
+testAccessResourceInsufficientScope :: Brig -> Nginz -> Http ()
+testAccessResourceInsufficientScope brig nginz = do
+  uid <- userId <$> createUser "alice" brig
+  let redirectUrl = mkUrl "https://example.com"
+  let scopes = OAuthScopes $ Set.fromList [WriteConversations]
+  (cid, secret, code) <- generateOAuthClientAndAuthorizationCode brig uid scopes redirectUrl
+  let accessTokenRequest = OAuthAccessTokenRequest OAuthGrantTypeAuthorizationCode cid secret code redirectUrl
+  accessToken <- createOAuthAccessToken brig accessTokenRequest
```

Comment:
this could also be another helper function taking `scopes` as an argument, making the tests yet a bit more concise.

Reply:
We need the url again further down, and sometimes we need uid, too. I found this to be a good balance,